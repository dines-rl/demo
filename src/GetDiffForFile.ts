import OpenAI from "openai";
const openai = new OpenAI();

import { zodResponseFormat } from "openai/helpers/zod";
import { z } from "zod";

const Change = z.object({
  lineStart: z.number({
    description:
      "The starting line number of the change in the original code file.",
  }),
  shortDescription: z.string({
    description: "A short description of the change. Suitable for a title",
  }),
  longDescription: z.string({
    description: "A longer description of the change. Suitable for a body",
  }),
  oldCode: z.string({ description: "The original code (before the change)." }),
  newCode: z.string({ description: "The updated code (after the change)." }),
});

const CodeSuggestionsResult = z.object({
  changes: z.array(Change, { description: "The changes made to the code." }),
  changedFileContents: z
    .string({
      description:
        "The code representing the entire changed file after the changes have been made.",
    })
    .optional(),
});

type Changes = {
  lineStart: number;
  shortDescription: string;
  longDescription: string;
  oldCode: string;
  newCode: string;
};

type GPTResultingData = {
  changes: Changes[];
  changed?: string;
  filename: string;
};

// Get code improvements from GPT
export async function getSuggestionsFromGPT(
  filename: string,
  code: string,
  {
    model = "gpt-4o-2024-08-06",
    temperature = 0.5,
    max_tokens = 10000, // TODo this is a lot
  }: { model?: string; temperature?: number; max_tokens?: number }
): Promise<GPTResultingData> {
  try {
    const completion = await openai.beta.chat.completions.parse({
      model: model,
      response_format: zodResponseFormat(CodeSuggestionsResult, "result"),
      messages: [
        {
          role: "system",
          content: `You are a code assistant tasked with improving the quality of the provided code. You will receive a complete code file, and your job is to review it, suggest improvements, and output the changes. You are a senior developer with extensive experience and also very helpful in the reasoning behind your changes.

### Instructions:
1. Analyze the provided code.
2. Suggest improvements and apply the changes.

Make sure to output both files as requested.
\`\`\`

`,
        },
        { role: "user", content: `\`\`\`${code}\`\`\`` },
      ],

      max_tokens: max_tokens,
      temperature: temperature,
    });

    const suggestion = completion.choices[0].message.parsed;
    console.log("--------------------");
    console.log("Suggestions from GPT:", suggestion);
    console.log("--------------------");

    if (suggestion) {
      return { filename, ...suggestion };
    } else {
      throw new Error("No suggestions received from GPT");
    }
  } catch (error) {
    console.error("Error fetching suggestions from GPT:", error);
    throw error;
  }
}
