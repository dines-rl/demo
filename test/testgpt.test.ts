import { describe, expect, test } from "vitest";
import { getSuggestionsFromGPT } from "../dist/GetDiffForFile.js";
import fs from "fs";
describe("GPTTest", () => {
  test("extracts the file changes and reasoning from the ai", async () => {
    const filename = "test/fixtures/testJSTFile.ts";
    const file = fs.readFileSync(filename, "utf-8");
    const result = await getSuggestionsFromGPT(filename, file, {
      temperature: 0.5,
      max_tokens: 500,
    });
    expect(result).toMatchSnapshot();
  });
});
