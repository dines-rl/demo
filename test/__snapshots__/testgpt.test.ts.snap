// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GPTTest > extracts the file changes and reasoning from the ai 1`] = `
{
  "changed": "const GLOBAL_VALUE = 1; // Changed to uppercase for constant naming convention
function addNumbers(a: number, b: number): number { // Renamed for clarity
  return a + b;
}",
  "changes": [
    {
      "Line End": 1,
      "Line Start": 1,
      "Long Description": "Changed the variable name 'globalValue' to 'GLOBAL_VALUE' to follow the convention of using uppercase for constants.",
      "New Code": "const GLOBAL_VALUE = 1;",
      "Old Code": "const globalValue = 1;",
      "Short Description": "Renamed constant to follow naming conventions",
    },
    {
      "Line End": 2,
      "Line Start": 2,
      "Long Description": "Changed the function name from 'MyFunction' to 'addNumbers' to better describe its purpose.",
      "New Code": "function addNumbers(a: number, b: number): number {",
      "Old Code": "function MyFunction(a: number, b: number): number {",
      "Short Description": "Renamed function for clarity",
    },
  ],
  "filename": "test/fixtures/testJSTFile.ts",
  "original": "\`\`\`changed
const GLOBAL_VALUE = 1; // Changed to uppercase for constant naming convention
function addNumbers(a: number, b: number): number { // Renamed for clarity
  return a + b;
}
\`\`\`

\`\`\`changes.json
[
    {
        "Line Start": 1,
        "Line End": 1,
        "Short Description": "Renamed constant to follow naming conventions",
        "Long Description": "Changed the variable name 'globalValue' to 'GLOBAL_VALUE' to follow the convention of using uppercase for constants.",
        "Old Code": "const globalValue = 1;",
        "New Code": "const GLOBAL_VALUE = 1;"
    },
    {
        "Line Start": 2,
        "Line End": 2,
        "Short Description": "Renamed function for clarity",
        "Long Description": "Changed the function name from 'MyFunction' to 'addNumbers' to better describe its purpose.",
        "Old Code": "function MyFunction(a: number, b: number): number {",
        "New Code": "function addNumbers(a: number, b: number): number {"
    }
]
\`\`\`",
}
`;

exports[`GPTTest > extracts the file changes and reasoning from the ai 2`] = `
{
  "changed": "const GLOBAL_VALUE = 1; // Changed to uppercase to follow constant naming conventions
function myFunction(a: number, b: number): number { // Changed function name to camelCase
  return a + b;
}",
  "changes": [
    {
      "Line End": 1,
      "Line Start": 1,
      "Long Description": "Changed the naming of the constant 'globalValue' to 'GLOBAL_VALUE' to follow the convention of using uppercase letters for constants.",
      "New Code": "const GLOBAL_VALUE = 1;",
      "Old Code": "const globalValue = 1;",
      "Short Description": "Renamed constant to uppercase",
    },
    {
      "Line End": 2,
      "Line Start": 2,
      "Long Description": "Changed the function name from 'MyFunction' to 'myFunction' to follow the JavaScript naming conventions for functions, which typically use camelCase.",
      "New Code": "function myFunction(a: number, b: number): number {",
      "Old Code": "function MyFunction(a: number, b: number): number {",
      "Short Description": "Renamed function to camelCase",
    },
  ],
  "filename": "test/fixtures/testJSTFile.ts",
  "original": "\`\`\`changed
const GLOBAL_VALUE = 1; // Changed to uppercase to follow constant naming conventions
function myFunction(a: number, b: number): number { // Changed function name to camelCase
  return a + b;
}
\`\`\`

\`\`\`changes.json
[
    {
        "Line Start": 1,
        "Line End": 1,
        "Short Description": "Renamed constant to uppercase",
        "Long Description": "Changed the naming of the constant 'globalValue' to 'GLOBAL_VALUE' to follow the convention of using uppercase letters for constants.",
        "Old Code": "const globalValue = 1;",
        "New Code": "const GLOBAL_VALUE = 1;"
    },
    {
        "Line Start": 2,
        "Line End": 2,
        "Short Description": "Renamed function to camelCase",
        "Long Description": "Changed the function name from 'MyFunction' to 'myFunction' to follow the JavaScript naming conventions for functions, which typically use camelCase.",
        "Old Code": "function MyFunction(a: number, b: number): number {",
        "New Code": "function myFunction(a: number, b: number): number {"
    }
]
\`\`\`",
}
`;
